client/server/user-bootstrap/test/index.js

```diff
@@ -0,0 +1,361 @@
+import { IncomingMessage } from 'http';
+import config from '@automattic/calypso-config';
+import nock from 'nock';
+import getBootstrappedUser from '../index';
+
+jest.mock( '@automattic/calypso-config', () => {
+	const impl = jest.fn();
+	impl.isEnabled = jest.fn();
+	return impl;
+} );
+
+const mockRequest = ( { cookies = {}, headers = {} } = {} ) => {
+	return Object.assign( new IncomingMessage(), {
+		body: {},
+		cookies,
+		query: {},
+		params: {},
+		get: jest.fn().mockImplementation( ( key ) => headers[ key ] ),
+	} );
+};
+
+const configureUpstreamRequest = ( {
+	headers = {},
+	response = { ID: '1' },
+	respondWithError = false,
+} = {} ) => {
+	return nock( 'https://public-api.wordpress.com', {
+		reqheaders: headers,
+	} )
+		.get( '/rest/v1/me?meta=flags' )
+		.reply( respondWithError ? 500 : 200, response );
+};
+
+const withConfig = ( keys ) => {
+	config.mockImplementation( ( key ) => keys[ key ] );
+};
+
+const withEnv = ( env = {} ) => {
+	for ( const [ k, v ] of Object.entries( env ) ) {
+		process.env[ k ] = v;
+	}
+};
+
+describe( 'User bootstrap', () => {
+	const realProcessEnv = { ...process.env };
+
+	beforeEach( () => {
+		// Default value for most tests. Some tests will overwrite this to simulate missing key
+		withConfig( {
+			wpcom_calypso_rest_api_key: 'key',
+		} );
+	} );
+
+	afterEach( () => {
+		nock.cleanAll();
+		jest.resetAllMocks();
+		for ( const [ env, value ] of Object.entries( realProcessEnv ) ) {
+			process.env[ env ] = value;
+		}
+	} );
+
+	it( 'throws if there is no auth cookie', async () => {
+		const request = mockRequest( { cookies: {} } );
+
+		await expect( getBootstrappedUser( request ) ).rejects.toThrow(
+			new Error( 'Cannot bootstrap without an auth cookie' )
+		);
+	} );
+
+	it( 'throws if there is supportSessionHeader and supportSessionCookie', async () => {
+		const request = mockRequest( {
+			cookies: {
+				wordpress_logged_in: 'auth-cookie',
+				support_session_id: 'session-id',
+			},
+			headers: {
+				'x-support-session': 'session-id',
+			},
+		} );
+
+		await expect( getBootstrappedUser( request ) ).rejects.toThrow(
+			new Error( 'Cannot bootstrap with both a support session header and support session cookie.' )
+		);
+	} );
+
+	it( 'sets the user agent in the upstream request', async () => {
+		const upstreamRequest = configureUpstreamRequest( {
+			headers: {
+				'User-Agent': 'WordPress.com Calypso',
+			},
+		} );
+		const request = mockRequest( {
+			cookies: {
+				wordpress_logged_in: 'auth-cookie',
+			},
+		} );
+
+		await getBootstrappedUser( request );
+
+		expect( upstreamRequest.isDone() ).toBe( true );
+	} );
+
+	it( 'sets the geo country code in the upstream request', async () => {
```

Comment:
Is this test really testing that the `x-geoip-country-code` is being added to the upstream request? It's not immediately clear. If it wasn't added, the `nock` mock, configured with `configureUpstreamRequest`, wouldn't match the request and the test would fail? Is this how `nock( { reqheaders } )` works?

Reply:
Yes, that's how it works. We configure an "expectation" in nock, it intercepts the calls and when we call `.isDone()`, it will return true if all expecations have been met.

I don't particuarly like that API, it doesn't play well with the semantics of `expect` or the pattern of arrange/act/assert that we use pretty much everywhere else. But AFAIK is the only way to use `nock` and I rather not introduce another different network mocking library.

Reply:
Using `nock` is probably fine, it's just hard to figure out what the test really does. Maybe some explanation comments could help?

> it intercepts the calls and when we call `.isDone()`

This is one things that's very unclear. When I do:
```js
nock( 'https://public-api.wordpress.com', { reqheaders: { 'Foo': 'bar' } } )
  .get( '/rest/v1/me?meta=flags' )
  .reply( { ID: '1' } )
```
and then send a request _without_ the `Foo` header, what happens? Does this mock match anyway and return `{ ID: '1' }`, and merely set `isDone` to `false` because the header was missing? Or will it fail to match at all and will cause the request to fail completely?

Reply:
Interesting! I thought it will not return anything and set `isDone` to `false`, but that's not how it works. It throws an error if we send a request that doesn't match any of the `nock` configurations.

So I can change all these tests: instead of using `isDone()` I can just assert that `getBootstrappedUser` resolves.